### ISSUE TITLE
generic method return types

### ISSUE THREAD

**Issue by ekuleshov** [dart-lang/language#2706]
The Dart language allows to specify specify generic type in the method parameters. So, this code works:

```
int useComparable<T extends Comparable>(T value, T value1) {
  return value.compareTo(value1);
}
```

The `T` is correctly recognized as a `Comparable` type.

However similar code does not work when `T` is used as a return value. E.g.

```
T getComparable<T extends Comparable>(dynamic value) {
  if(value is int) return value;
  return value.toString();
}
```

or 

```
T getComparable<T extends Comparable>(T value) {
  if(value is int) return value;
  return value.toString();
}
```

These fail at compile time with the following error:
* A value of type 'int' can't be returned from the function 'getComparable' because it has a return type of 'T'.
* A value of type 'String' can't be returned from the function 'getComparable' because it has a return type of 'T'.

I could get around this by adding explicit cast:

```
T getComparable<T extends Comparable>(dynamic value) {
  if(value is int) return value as T;
  return value.toString() as T;
}
```

But that should not be necessary as T is declared to extend Comparable type. So, it would be good if compiler would be able to recognize declared generic type without that.

**Comment by lrhn** [comment]
This is working as intended.

In the first example, you do not specify the type argument to `Comparable`, so it defaults to `Object?`, becoming effectively:
```dart
int useComparable<T extends Comparable<Object?>>(T value, T value1) {
  return value.compareTo(value1);
}
```
The code runs, because `value` being `Comparable<Object>` means that it allows you to call `value.compareTo(...)` with any argument.

The second example is likewise equivalent to:
```dart
T getComparable<T extends Comparable<Object?>>(dynamic value) {
  if(value is int) return value;
  return value.toString();
}
```
However, you have to return something which is a `T`, not just a `Comparable<Object?>`, and the actual type of `T` may be any subtype of `Comparable<T>`. Including `Never`. The compiler cannot know whether `int` is a subtype of `T`, so the program is not allowed.

You might just want to have a return type of `Comparable`, not a type variable which extends `Comparable` in an unpredictable (to the compiler) way.

In short, `T` extends `Comparable` (aka. `Comparable<Object?>`), but that does not mean that every `Comparable` extends `T`. Rather the opposite. The compiler errors are correct.
